### DISCUSSION TITLE
Media and Exchange Metabolite Elasticities

### DISCUSSION THREAD

**Discussion by ezgitas** [micom-dev/micom#52] [Q&A] [answered]
Hi! 

I have two unrelated problems: one is about gut media implantation and the other is about elasticities.

 I am trying to implement low carb - high fat diet media to my community model and using the media that you shared. I followed the same steps for western diet media integration and it worked perfectly.

`western_medium = load_qiime_medium("./media/media/western_diet_gut_agora.qza")`
`medium = western_medium`
`communityFile = path + sampleName + '.pickle'`
`com = micom.load_pickle(communityFile)`
`completed = complete_medium(com, medium.flux, growth=0.258).to_frame('flux')`
`completed["reaction"] = completed.index`
`com.medium = completed.flux`

When I define medium as low carb - high fat diet the following error appears:

`[20:53:52] WARNING  solver encountered an error infeasible                 solution.py:220`
`Traceback (most recent call last):`
`File "cooperative_tradeoff.py", line 51, in <module>`
` completed = complete_medium(com, medium.flux, growth=0.258).to_frame('flux')`
`File "/opt/anaconda3/envs/micom/lib/python3.8/site-packages/micom/media.py", line 351, in complete_medium`
`fluxes = sol.fluxes.loc["medium", :]`
`AttributeError: 'NoneType' object has no attribute 'fluxes'`

The elasticity problem is about exchange metabolites. I can calculate elasticities for exchange reactions and abundances but is there a way to calculate elasticities for exchange metabolites? 

Lastly, one little question. What is the difference between `elasticities_by_medium` and `elasticities`? The elasticity values are different and one uses only metabolite while the other uses both metabolite and taxa, I know these but could not figure out the main difference.

Thanks in advance.
Ezgi

**Comment by cdiener** [comment, marked as answer]
Hi, the first error means that the model can't reach the specified growth rate with the given diet (that's what the warning says). `complete_medium` adds in additional components to the ones already set in the medium to achieve the specified growth rate. Sometimes the template medium is already too restrictive to allow for large growth rates. The low carb medium has exactly this restriction. Since it restricts most carbon sources this usually only allows very small growth rates. For instance, in the diagnostics [run in the media construction](https://nbviewer.jupyter.org/github/micom-dev/media/blob/main/recipes/vmh_high_fat_low_carb_agora.ipynb) the largest observed growth rate for any single organism was 0.006 compared to 0.5 for the western medium. So you would have to lower the requested growth rate for that too work. You could go around that by relaxing the flux constraints in the template (the low carb diet) by either (1) raising very small fluxes with something like `medium.loc[medium.flux < 1e-3, "flux"] = 1e-3` (or something similar) or (2) just raising all flux constraints with `medium.flux *= 1000` before you run complete_medium. Also note that the medium provided on the [media repo](https://github.com/micom-dev/media) is already completed for all individual genera in AGORA. So they can all grow on that medium and you could just use it in `grow()` as is.  

FBA does not model metabolite concentrations, so the closest thing to getting the elasticity for a metabolite concentration is to use the import flux which is the exchange reaction. Those model how much of a metabolite is supplied into the system. 

`elasticities_by_medium` and `elasticities_by_abundance` are just helper functions that are used by elasticities to calculate the elasticities for the exchange reactions and taxa abundances respectively.
